receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+			if clientSNMPData.valueType == notSupportedVal {
+				scraperErrors.AddPartial(1, fmt.Errorf("problem with getting indexed data: data for OID '%s' not a supported type", data.Name))
+				return nil
+			}
+
+			// Process the data
+			if err := processFn(clientSNMPData); err != nil {
+				scraperErrors.AddPartial(1, fmt.Errorf("problem with getting indexed data: problem with processing data for OID '%s': %w", data.Name, err))
+			}
+
+			return nil
+		}
+
+		// Call the correct gosnmp Walk function based on SNMP version
+		var err error
+		if c.client.GetVersion() == gosnmp.Version1 {
+			err = c.client.Walk(oid, walkFn)
+		} else {
+			err = c.client.BulkWalk(oid, walkFn)
+		}
+		if err != nil {
+			scraperErrors.AddPartial(1, fmt.Errorf("problem with getting indexed data: problem with SNMP WALK for OID '%v': %w", oid, err))
+			// Allows for quicker recovery rather than timing out for each WALK OID and waiting for the next GET to fix it
+			if strings.Contains(err.Error(), "request timeout (after ") {
+				if err = c.Close(); err != nil {
+					c.logger.Warn("Problem with closing connection while trying to reset it", zap.Error(err))
+				}
+				if err = c.Connect(); err != nil {
+					scraperErrors.AddPartial(len(oids), fmt.Errorf("problem with getting indexed data: problem connecting while trying to reset connection: %w", err))
+					return
+				}
+			}
+		} else if !walkFnCalled {
+			scraperErrors.AddPartial(1, fmt.Errorf("problem with getting indexed data: problem with SNMP WALK for OID '%v': Could not find any data for given OID", oid))
+		}
+	}
+}
+
+// chunkArray takes an initial array and splits it into a number of smaller
+// arrays of a given size.
+func chunkArray(initArray []string, chunkSize int) [][]string {
+	var chunkedArrays [][]string
+
+	for i := 0; i < len(initArray); i += chunkSize {
+		end := i + chunkSize
+
+		if end > len(initArray) {
+			end = len(initArray)
+		}
+
+		chunkedArrays = append(chunkedArrays, initArray[i:end])
+	}
+	return chunkedArrays
+}
+
+// convertSnmpPDUToSnmpData takes a piece of SnmpPDU data and converts it to the
+// client's snmpData type.
+func (c *snmpClient) convertSnmpPDUToSnmpData(pdu gosnmp.SnmpPDU) snmpData {
+	clientSNMPData := snmpData{
+		oid: pdu.Name,
+	}
+
+	// Condense gosnmp data types to our client's simplified data types
+	switch pdu.Type {
+	// Integer types
+	case gosnmp.Counter32:
+		fallthrough
+	case gosnmp.Gauge32:
+		fallthrough
+	case gosnmp.Uinteger32:
+		fallthrough
+	case gosnmp.TimeTicks:
+		fallthrough
+	case gosnmp.Integer:
+		value, err := c.toInt64(pdu.Name, pdu.Value)
+		if err != nil {
+			clientSNMPData.valueType = notSupportedVal
+			clientSNMPData.value = value
+			return clientSNMPData
+		}
+
+		clientSNMPData.valueType = integerVal
+		clientSNMPData.value = value
+		return clientSNMPData
+
+	// String types
+	case gosnmp.IPAddress:
+		fallthrough
+	case gosnmp.ObjectIdentifier:
+		fallthrough
+	case gosnmp.OctetString:
```

Comment:
Up to you, but in my opinion, this is a lot easier to read. 
```suggestion
	case gosnmp.IPAddress, gosnmp.ObjectIdentifier, gosnmp.OctetString:
```

`fallthrough` is useful when you actually need to differentiate cases but also want to share a part of the handling. If the handling is exactly the same, I think it's better to just group them.

Reply:
@djaglowski i will switch this. chalk it up to forgetting my options with go